cub/cub/device/dispatch/dispatch_transform.cuh

```diff
@@ -0,0 +1,1277 @@
+  RandomAccessIteratorIn... ins)
+{
+  constexpr int block_dim = PrefetchPolicy::BLOCK_THREADS;
+  {
+    const int tile_stride = block_dim * num_elem_per_thread;
+    const Offset offset   = static_cast<Offset>(blockIdx.x) * tile_stride;
+
+    // move index and iterator domain to the block/thread index, to reduce arithmetic in the loops below
+    assert(offset < num_items);
+    num_items -= offset;
+    int dummy[] = {(ins += offset, 0)..., 0};
+    (void) &dummy;
+    out += offset;
+  }
+
+  for (int j = 0; j < num_elem_per_thread; ++j)
+  {
+    const int idx = j * block_dim + threadIdx.x;
+    // TODO(bgruber): replace by fold over comma in C++17
+    int dummy[] = {(prefetch(ins + idx), 0)..., 0}; // extra zero to handle empty packs
+    (void) &dummy; // nvcc 11.1 needs extra strong unused warning suppression
+    (void) &idx; // nvcc 11.1 needs extra strong unused warning suppression
+  }
+
+  // ahendriksen: various unrolling yields less <1% gains at much higher compile-time cost
+  // TODO(bgruber): A6000 disagrees
+#pragma unroll 1
+  for (int j = 0; j < num_elem_per_thread; ++j)
+  {
+    const int idx = j * block_dim + threadIdx.x;
+    if (idx < num_items)
+    {
+      // we have to unwrap Thrust's proxy references here for backward compatibility (try zip_iterator.cu test)
+      out[idx] = f(THRUST_NS_QUALIFIER::raw_reference_cast(ins[idx])...);
+    }
+  }
+}
+
+template <int BlockThreads, int ItemsPerThread>
+struct unrolled_policy_t
+{
+  static constexpr int BLOCK_THREADS    = BlockThreads;
+  static constexpr int ITEMS_PER_THREAD = ItemsPerThread;
+};
+
+// ahendriksen: no __restrict__ should be necessary on the input pointers since we already separated the load stage from
+// the store stage.
+template <typename UnrolledPolicy,
+          typename Offset,
+          typename F,
+          typename RandomAccessIteratorOut,
+          typename... RandomAccessIteratorIn>
+_CCCL_DEVICE void transform_kernel_impl(
+  ::cuda::std::integral_constant<Algorithm, Algorithm::unrolled_staged>,
+  Offset num_items,
+  int,
+  F f,
+  RandomAccessIteratorOut out,
+  RandomAccessIteratorIn... ins)
+{
+  constexpr int block_dim        = UnrolledPolicy::BLOCK_THREADS;
+  constexpr int items_per_thread = UnrolledPolicy::ITEMS_PER_THREAD;
+
+  {
+    constexpr int tile_stride = block_dim * items_per_thread;
+    const Offset offset       = static_cast<Offset>(blockIdx.x) * tile_stride;
+
+    // move index and iterator domain to the block/thread index, to reduce arithmetic in the loops below
+    num_items -= offset;
+    int dummy[] = {(ins += offset, 0)..., 0};
+    (void) dummy;
+    out += offset;
+  }
+
+  // TODO(bgruber): we could use load vectorization here
+
+  [&](cuda::std::array<value_t<RandomAccessIteratorIn>, items_per_thread>&&... arrays) {
+  // load items_per_thread elements
+#pragma unroll
+    for (int j = 0; j < items_per_thread; ++j)
+    {
+      const int idx = j * block_dim + threadIdx.x;
+      if (idx < num_items)
```

Comment:
Full tiles should skip this check. The compiler is not able to determine the range of threadIdx.x .

Reply:
I was initially hesitant of this suggestion, since it puts the user function object into the code twice. I also thought that this branch cannot be so costly. I implemented and benchmarked it now, and boy was I wrong.

This change has the biggest impact on kernels with small bytes needed per tile, e.g. Babelstream mul, which only reads 1 item per computation. Especially on small data types (`char` and `short`) there is a lot of compute in this loop. Because the BIF maximiation on the host tries to scale up the input tile size to reach a certain BIF per SM, the tile to compute gets super large for small data types, so each instruction we can get out of this loop really matters. Here is some data from H200:

```
# mul

## [0] NVIDIA H200

|  T{ct}  |  OffsetT{ct}  |  Elements{io}  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |       Diff |   %Diff |  Status  |
|---------|---------------|----------------|------------|-------------|------------|-------------|------------|---------|----------|
|   I8    |      I32      |      2^16      |   4.987 us |       3.13% |   4.359 us |       3.62% |  -0.628 us | -12.59% |   FAIL   |
|   I8    |      I32      |      2^20      |   5.893 us |       3.15% |   5.262 us |       3.74% |  -0.631 us | -10.71% |   FAIL   |
|   I8    |      I32      |      2^24      |  15.899 us |       1.67% |  14.007 us |       2.34% |  -1.891 us | -11.90% |   FAIL   |
|   I8    |      I32      |      2^28      | 158.766 us |       0.47% | 137.215 us |       0.65% | -21.551 us | -13.57% |   FAIL   |
|   I8    |      I64      |      2^16      |   5.372 us |       2.82% |   4.848 us |       2.93% |  -0.523 us |  -9.74% |   FAIL   |
|   I8    |      I64      |      2^20      |   5.916 us |       2.32% |   5.373 us |       2.88% |  -0.544 us |  -9.19% |   FAIL   |
|   I8    |      I64      |      2^24      |  15.766 us |       1.67% |  14.080 us |       2.51% |  -1.685 us | -10.69% |   FAIL   |
|   I8    |      I64      |      2^28      | 154.831 us |       0.55% | 137.225 us |       0.68% | -17.606 us | -11.37% |   FAIL   |
|   I16   |      I32      |      2^16      |   4.954 us |       2.71% |   4.691 us |       3.05% |  -0.263 us |  -5.32% |   FAIL   |
|   I16   |      I32      |      2^20      |   6.098 us |       2.89% |   5.685 us |       2.81% |  -0.413 us |  -6.78% |   FAIL   |
|   I16   |      I32      |      2^24      |  22.835 us |       1.92% |  21.539 us |       2.23% |  -1.296 us |  -5.68% |   FAIL   |
|   I16   |      I32      |      2^28      | 278.136 us |       0.23% | 254.731 us |       0.37% | -23.405 us |  -8.41% |   FAIL   |
|   I16   |      I64      |      2^16      |   4.986 us |       2.62% |   4.700 us |       2.86% |  -0.286 us |  -5.73% |   FAIL   |
|   I16   |      I64      |      2^20      |   6.148 us |       2.85% |   5.759 us |       2.80% |  -0.390 us |  -6.34% |   FAIL   |
|   I16   |      I64      |      2^24      |  22.810 us |       1.80% |  21.625 us |       2.41% |  -1.185 us |  -5.20% |   FAIL   |
|   I16   |      I64      |      2^28      | 279.849 us |       0.21% | 255.242 us |       0.38% | -24.606 us |  -8.79% |   FAIL   |
|   F32   |      I32      |      2^16      |   4.815 us |       2.81% |   4.739 us |       2.93% |  -0.076 us |  -1.58% |   PASS   |
|   F32   |      I32      |      2^20      |   6.835 us |       3.25% |   6.811 us |       3.18% |  -0.025 us |  -0.36% |   PASS   |
|   F32   |      I32      |      2^24      |  37.069 us |       1.95% |  37.052 us |       1.86% |  -0.017 us |  -0.05% |   PASS   |
|   F32   |      I32      |      2^28      | 508.218 us |       0.19% | 509.450 us |       0.19% |   1.232 us |   0.24% |   FAIL   |
|   F32   |      I64      |      2^16      |   4.873 us |       3.83% |   4.838 us |       4.11% |  -0.035 us |  -0.72% |   PASS   |
|   F32   |      I64      |      2^20      |   6.770 us |       2.78% |   6.726 us |       2.83% |  -0.045 us |  -0.66% |   PASS   |
|   F32   |      I64      |      2^24      |  36.925 us |       1.88% |  36.982 us |       1.85% |   0.057 us |   0.16% |   PASS   |
|   F32   |      I64      |      2^28      | 507.866 us |       0.20% | 509.104 us |       0.19% |   1.238 us |   0.24% |   FAIL   |
|   F64   |      I32      |      2^16      |   4.877 us |       3.25% |   4.911 us |       3.75% |   0.034 us |   0.70% |   PASS   |
|   F64   |      I32      |      2^20      |   8.989 us |       3.11% |   9.006 us |       2.95% |   0.017 us |   0.19% |   PASS   |
|   F64   |      I32      |      2^24      |  67.594 us |       1.17% |  67.596 us |       1.12% |   0.002 us |   0.00% |   PASS   |
|   F64   |      I32      |      2^28      |   1.011 ms |       0.12% |   1.011 ms |       0.12% |   0.353 us |   0.03% |   PASS   |
|   F64   |      I64      |      2^16      |   4.899 us |       3.47% |   4.964 us |       4.39% |   0.065 us |   1.32% |   PASS   |
|   F64   |      I64      |      2^20      |   8.910 us |       2.67% |   8.978 us |       3.01% |   0.068 us |   0.76% |   PASS   |
|   F64   |      I64      |      2^24      |  67.496 us |       1.17% |  67.525 us |       1.18% |   0.029 us |   0.04% |   PASS   |
|   F64   |      I64      |      2^28      |   1.011 ms |       0.13% |   1.011 ms |       0.13% |  -0.154 us |  -0.02% |   PASS   |
|  I128   |      I32      |      2^16      |   5.234 us |       4.63% |   5.282 us |       4.71% |   0.048 us |   0.92% |   PASS   |
|  I128   |      I32      |      2^20      |  13.224 us |       2.25% |  13.278 us |       2.73% |   0.054 us |   0.41% |   PASS   |
|  I128   |      I32      |      2^24      | 133.157 us |       0.66% | 133.202 us |       0.67% |   0.045 us |   0.03% |   PASS   |
|  I128   |      I32      |      2^28      |   2.052 ms |       0.08% |   2.052 ms |       0.08% |   0.229 us |   0.01% |   PASS   |
|  I128   |      I64      |      2^16      |   4.979 us |       3.25% |   5.147 us |       3.35% |   0.168 us |   3.38% |   FAIL   |
|  I128   |      I64      |      2^20      |  13.072 us |       2.43% |  13.153 us |       2.46% |   0.081 us |   0.62% |   PASS   |
|  I128   |      I64      |      2^24      | 134.486 us |       0.68% | 134.548 us |       0.67% |   0.062 us |   0.05% |   PASS   |
|  I128   |      I64      |      2^28      |   2.054 ms |       0.36% |   2.055 ms |       0.33% |   1.035 us |   0.05% |   PASS   |

# add

## [0] NVIDIA H200

|  T{ct}  |  OffsetT{ct}  |  Elements{io}  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |       Diff |   %Diff |  Status  |
|---------|---------------|----------------|------------|-------------|------------|-------------|------------|---------|----------|
|   I8    |      I32      |      2^16      |   4.601 us |       2.97% |   4.368 us |       3.18% |  -0.233 us |  -5.06% |   FAIL   |
|   I8    |      I32      |      2^20      |   5.957 us |       3.68% |   5.709 us |       3.60% |  -0.248 us |  -4.17% |   FAIL   |
|   I8    |      I32      |      2^24      |  20.002 us |       1.79% |  18.961 us |       1.86% |  -1.041 us |  -5.21% |   FAIL   |
|   I8    |      I32      |      2^28      | 220.364 us |       0.28% | 208.517 us |       0.23% | -11.847 us |  -5.38% |   FAIL   |
|   I8    |      I64      |      2^16      |   5.077 us |       3.07% |   4.872 us |       2.79% |  -0.205 us |  -4.04% |   FAIL   |
|   I8    |      I64      |      2^20      |   6.002 us |       2.59% |   5.848 us |       3.68% |  -0.154 us |  -2.57% |   PASS   |
|   I8    |      I64      |      2^24      |  20.581 us |       1.58% |  18.946 us |       1.90% |  -1.635 us |  -7.95% |   FAIL   |
|   I8    |      I64      |      2^28      | 229.251 us |       0.28% | 209.389 us |       0.24% | -19.862 us |  -8.66% |   FAIL   |
|   I16   |      I32      |      2^16      |   4.843 us |       2.57% |   4.785 us |       2.65% |  -0.058 us |  -1.20% |   PASS   |
|   I16   |      I32      |      2^20      |   6.768 us |       2.87% |   6.582 us |       2.57% |  -0.186 us |  -2.75% |   FAIL   |
|   I16   |      I32      |      2^24      |  31.029 us |       1.67% |  30.748 us |       1.86% |  -0.281 us |  -0.91% |   PASS   |
|   I16   |      I32      |      2^28      | 401.222 us |       0.13% | 375.325 us |       0.22% | -25.896 us |  -6.45% |   FAIL   |
|   I16   |      I64      |      2^16      |   4.974 us |       3.15% |   4.849 us |       3.03% |  -0.125 us |  -2.52% |   PASS   |
|   I16   |      I64      |      2^20      |   6.850 us |       3.14% |   6.698 us |       3.19% |  -0.151 us |  -2.21% |   PASS   |
|   I16   |      I64      |      2^24      |  31.830 us |       1.83% |  30.876 us |       1.78% |  -0.953 us |  -3.00% |   FAIL   |
|   I16   |      I64      |      2^28      | 405.690 us |       0.11% | 379.833 us |       0.21% | -25.858 us |  -6.37% |   FAIL   |
|   F32   |      I32      |      2^16      |   4.917 us |       2.82% |   5.030 us |       3.70% |   0.113 us |   2.29% |   PASS   |
|   F32   |      I32      |      2^20      |   8.301 us |       2.80% |   8.214 us |       2.93% |  -0.087 us |  -1.04% |   PASS   |
|   F32   |      I32      |      2^24      |  53.176 us |       1.37% |  53.178 us |       1.36% |   0.003 us |   0.00% |   PASS   |
|   F32   |      I32      |      2^28      | 731.946 us |       0.11% | 732.111 us |       0.11% |   0.166 us |   0.02% |   PASS   |
|   F32   |      I64      |      2^16      |   4.960 us |       3.56% |   5.067 us |       3.03% |   0.107 us |   2.16% |   PASS   |
|   F32   |      I64      |      2^20      |   8.315 us |       2.83% |   8.237 us |       3.02% |  -0.078 us |  -0.94% |   PASS   |
|   F32   |      I64      |      2^24      |  53.274 us |       1.39% |  53.285 us |       1.38% |   0.011 us |   0.02% |   PASS   |
|   F32   |      I64      |      2^28      | 732.988 us |       0.11% | 732.308 us |       0.11% |  -0.679 us |  -0.09% |   PASS   |
|   F64   |      I32      |      2^16      |   5.307 us |       3.69% |   5.109 us |       3.98% |  -0.198 us |  -3.74% |   FAIL   |
|   F64   |      I32      |      2^20      |  11.558 us |       2.41% |  11.475 us |       2.69% |  -0.082 us |  -0.71% |   PASS   |
|   F64   |      I32      |      2^24      |  97.937 us |       0.51% |  97.827 us |       0.46% |  -0.110 us |  -0.11% |   PASS   |
|   F64   |      I32      |      2^28      |   1.466 ms |       0.11% |   1.466 ms |       0.11% |  -0.034 us |  -0.00% |   PASS   |
|   F64   |      I64      |      2^16      |   5.400 us |       4.13% |   5.209 us |       4.66% |  -0.191 us |  -3.54% |   PASS   |
|   F64   |      I64      |      2^20      |  11.576 us |       2.69% |  11.502 us |       2.57% |  -0.075 us |  -0.65% |   PASS   |
|   F64   |      I64      |      2^24      |  97.929 us |       0.53% |  97.881 us |       0.52% |  -0.048 us |  -0.05% |   PASS   |
|   F64   |      I64      |      2^28      |   1.466 ms |       0.11% |   1.466 ms |       0.11% |  -0.429 us |  -0.03% |   PASS   |
|  I128   |      I32      |      2^16      |   5.787 us |       4.10% |   5.602 us |       4.23% |  -0.185 us |  -3.20% |   PASS   |
|  I128   |      I32      |      2^20      |  17.660 us |       2.17% |  17.785 us |       2.24% |   0.125 us |   0.71% |   PASS   |
|  I128   |      I32      |      2^24      | 189.005 us |       0.35% | 188.794 us |       0.29% |  -0.211 us |  -0.11% |   PASS   |
|  I128   |      I32      |      2^28      |   2.930 ms |       0.06% |   2.907 ms |       0.10% | -23.307 us |  -0.80% |   FAIL   |
|  I128   |      I64      |      2^16      |   5.441 us |       3.17% |   5.473 us |       3.36% |   0.033 us |   0.60% |   PASS   |
|  I128   |      I64      |      2^20      |  17.733 us |       2.35% |  17.874 us |       2.21% |   0.141 us |   0.79% |   PASS   |
|  I128   |      I64      |      2^24      | 190.714 us |       0.32% | 190.650 us |       0.29% |  -0.064 us |  -0.03% |   PASS   |
|  I128   |      I64      |      2^28      |   2.934 ms |       0.44% |   2.907 ms |       0.48% | -27.268 us |  -0.93% |   FAIL   |

# triad

## [0] NVIDIA H200

|  T{ct}  |  OffsetT{ct}  |  Elements{io}  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |       Diff |   %Diff |  Status  |
|---------|---------------|----------------|------------|-------------|------------|-------------|------------|---------|----------|
|   I8    |      I32      |      2^16      |   5.145 us |       3.11% |   4.883 us |       3.46% |  -0.262 us |  -5.10% |   FAIL   |
|   I8    |      I32      |      2^20      |   5.937 us |       2.76% |   5.717 us |       3.41% |  -0.220 us |  -3.70% |   FAIL   |
|   I8    |      I32      |      2^24      |  20.110 us |       1.87% |  18.907 us |       2.09% |  -1.203 us |  -5.98% |   FAIL   |
|   I8    |      I32      |      2^28      | 221.547 us |       0.28% | 210.049 us |       0.28% | -11.498 us |  -5.19% |   FAIL   |
|   I8    |      I64      |      2^16      |   5.135 us |       3.13% |   4.911 us |       3.06% |  -0.224 us |  -4.36% |   FAIL   |
|   I8    |      I64      |      2^20      |   6.022 us |       3.28% |   5.763 us |       3.24% |  -0.259 us |  -4.31% |   FAIL   |
|   I8    |      I64      |      2^24      |  20.575 us |       1.68% |  18.813 us |       2.07% |  -1.762 us |  -8.57% |   FAIL   |
|   I8    |      I64      |      2^28      | 229.874 us |       0.29% | 210.178 us |       0.88% | -19.696 us |  -8.57% |   FAIL   |
|   I16   |      I32      |      2^16      |   4.859 us |       2.60% |   4.828 us |       2.66% |  -0.031 us |  -0.63% |   PASS   |
|   I16   |      I32      |      2^20      |   6.810 us |       2.78% |   6.630 us |       2.98% |  -0.180 us |  -2.64% |   PASS   |
|   I16   |      I32      |      2^24      |  31.050 us |       1.72% |  30.207 us |       1.97% |  -0.843 us |  -2.72% |   FAIL   |
|   I16   |      I32      |      2^28      | 394.250 us |       0.17% | 373.695 us |       0.21% | -20.555 us |  -5.21% |   FAIL   |
|   I16   |      I64      |      2^16      |   4.954 us |       2.94% |   4.875 us |       3.15% |  -0.079 us |  -1.59% |   PASS   |
|   I16   |      I64      |      2^20      |   6.878 us |       2.75% |   6.646 us |       3.00% |  -0.232 us |  -3.38% |   FAIL   |
|   I16   |      I64      |      2^24      |  31.388 us |       1.78% |  30.308 us |       1.78% |  -1.080 us |  -3.44% |   FAIL   |
|   I16   |      I64      |      2^28      | 401.670 us |       0.14% | 376.498 us |       0.21% | -25.172 us |  -6.27% |   FAIL   |
|   F32   |      I32      |      2^16      |   4.846 us |       4.08% |   4.866 us |       2.98% |   0.020 us |   0.41% |   PASS   |
|   F32   |      I32      |      2^20      |   8.178 us |       3.41% |   8.235 us |       2.54% |   0.057 us |   0.70% |   PASS   |
|   F32   |      I32      |      2^24      |  53.278 us |       1.37% |  53.389 us |       1.38% |   0.112 us |   0.21% |   PASS   |
|   F32   |      I32      |      2^28      | 731.805 us |       0.11% | 731.879 us |       0.10% |   0.074 us |   0.01% |   PASS   |
|   F32   |      I64      |      2^16      |   4.866 us |       4.31% |   4.978 us |       4.17% |   0.112 us |   2.30% |   PASS   |
|   F32   |      I64      |      2^20      |   8.164 us |       2.47% |   8.231 us |       2.81% |   0.067 us |   0.83% |   PASS   |
|   F32   |      I64      |      2^24      |  53.352 us |       1.37% |  53.365 us |       1.36% |   0.013 us |   0.02% |   PASS   |
|   F32   |      I64      |      2^28      | 732.253 us |       0.11% | 732.006 us |       0.11% |  -0.247 us |  -0.03% |   PASS   |
|   F64   |      I32      |      2^16      |   5.081 us |       3.28% |   5.203 us |       3.84% |   0.122 us |   2.41% |   PASS   |
|   F64   |      I32      |      2^20      |  11.560 us |       2.72% |  11.598 us |       2.87% |   0.038 us |   0.33% |   PASS   |
|   F64   |      I32      |      2^24      |  97.423 us |       0.58% |  97.529 us |       0.55% |   0.106 us |   0.11% |   PASS   |
|   F64   |      I32      |      2^28      |   1.467 ms |       0.10% |   1.466 ms |       0.10% |  -0.269 us |  -0.02% |   PASS   |
|   F64   |      I64      |      2^16      |   5.160 us |       4.15% |   5.321 us |       4.53% |   0.161 us |   3.12% |   PASS   |
|   F64   |      I64      |      2^20      |  11.499 us |       2.12% |  11.599 us |       2.63% |   0.100 us |   0.87% |   PASS   |
|   F64   |      I64      |      2^24      |  97.599 us |       0.45% |  97.530 us |       0.56% |  -0.069 us |  -0.07% |   PASS   |
|   F64   |      I64      |      2^28      |   1.467 ms |       0.10% |   1.466 ms |       0.10% |  -0.117 us |  -0.01% |   PASS   |
|  I128   |      I32      |      2^16      |   5.589 us |       4.15% |   5.720 us |       4.60% |   0.131 us |   2.34% |   PASS   |
|  I128   |      I32      |      2^20      |  17.868 us |       2.28% |  17.963 us |       2.17% |   0.096 us |   0.53% |   PASS   |
|  I128   |      I32      |      2^24      | 188.877 us |       0.32% | 188.883 us |       0.35% |   0.006 us |   0.00% |   PASS   |
|  I128   |      I32      |      2^28      |   2.933 ms |       0.06% |   2.933 ms |       0.06% |  -0.590 us |  -0.02% |   PASS   |
|  I128   |      I64      |      2^16      |   5.427 us |       3.57% |   5.562 us |       3.42% |   0.135 us |   2.48% |   PASS   |
|  I128   |      I64      |      2^20      |  17.837 us |       2.36% |  17.959 us |       2.30% |   0.122 us |   0.68% |   PASS   |
|  I128   |      I64      |      2^24      | 191.012 us |       0.33% | 191.097 us |       0.32% |   0.085 us |   0.04% |   PASS   |
|  I128   |      I64      |      2^28      |   2.933 ms |       0.43% |   2.932 ms |       0.35% |  -0.597 us |  -0.02% |   PASS   |

# nstream

## [0] NVIDIA H200

|  T{ct}  |  OffsetT{ct}  |  Elements{io}  |  OverwriteInput  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |       Diff |   %Diff |  Status  |
|---------|---------------|----------------|------------------|------------|-------------|------------|-------------|------------|---------|----------|
|   I8    |      I32      |      2^16      |        1         |   4.678 us |       3.39% |   4.589 us |       3.09% |  -0.089 us |  -1.90% |   PASS   |
|   I8    |      I32      |      2^20      |        1         |   6.229 us |       2.66% |   6.211 us |       2.71% |  -0.018 us |  -0.28% |   PASS   |
|   I8    |      I32      |      2^24      |        1         |  24.048 us |       1.97% |  23.639 us |       1.86% |  -0.409 us |  -1.70% |   PASS   |
|   I8    |      I32      |      2^28      |        1         | 286.454 us |       0.21% | 284.845 us |       0.20% |  -1.608 us |  -0.56% |   FAIL   |
|   I8    |      I64      |      2^16      |        1         |   5.215 us |       3.06% |   5.122 us |       3.13% |  -0.093 us |  -1.77% |   PASS   |
|   I8    |      I64      |      2^20      |        1         |   6.297 us |       2.18% |   6.261 us |       3.18% |  -0.037 us |  -0.58% |   PASS   |
|   I8    |      I64      |      2^24      |        1         |  24.893 us |       1.81% |  23.674 us |       1.97% |  -1.219 us |  -4.90% |   FAIL   |
|   I8    |      I64      |      2^28      |        1         | 298.657 us |       0.16% | 286.044 us |       0.20% | -12.613 us |  -4.22% |   FAIL   |
|   I16   |      I32      |      2^16      |        1         |   4.928 us |       2.97% |   4.975 us |       3.19% |   0.047 us |   0.95% |   PASS   |
|   I16   |      I32      |      2^20      |        1         |   7.445 us |       2.94% |   7.404 us |       2.85% |  -0.040 us |  -0.54% |   PASS   |
|   I16   |      I32      |      2^24      |        1         |  40.062 us |       1.67% |  38.559 us |       1.71% |  -1.502 us |  -3.75% |   FAIL   |
|   I16   |      I32      |      2^28      |        1         | 538.376 us |       0.16% | 494.007 us |       0.10% | -44.369 us |  -8.24% |   FAIL   |
|   I16   |      I64      |      2^16      |        1         |   4.969 us |       3.29% |   5.030 us |       3.62% |   0.061 us |   1.24% |   PASS   |
|   I16   |      I64      |      2^20      |        1         |   7.456 us |       2.76% |   7.447 us |       2.70% |  -0.009 us |  -0.12% |   PASS   |
|   I16   |      I64      |      2^24      |        1         |  40.041 us |       1.65% |  38.870 us |       1.68% |  -1.171 us |  -2.92% |   FAIL   |
|   I16   |      I64      |      2^28      |        1         | 538.248 us |       0.16% | 504.101 us |       0.11% | -34.147 us |  -6.34% |   FAIL   |
|   F32   |      I32      |      2^16      |        1         |   5.004 us |       3.62% |   5.024 us |       2.98% |   0.019 us |   0.39% |   PASS   |
|   F32   |      I32      |      2^20      |        1         |   9.600 us |       2.76% |   9.627 us |       2.77% |   0.027 us |   0.29% |   PASS   |
|   F32   |      I32      |      2^24      |        1         |  68.613 us |       1.02% |  68.792 us |       1.03% |   0.178 us |   0.26% |   PASS   |
|   F32   |      I32      |      2^28      |        1         | 971.161 us |       0.09% | 973.359 us |       0.08% |   2.197 us |   0.23% |   FAIL   |
|   F32   |      I64      |      2^16      |        1         |   5.065 us |       4.15% |   5.082 us |       4.99% |   0.017 us |   0.33% |   PASS   |
|   F32   |      I64      |      2^20      |        1         |   9.603 us |       2.59% |   9.600 us |       2.56% |  -0.003 us |  -0.03% |   PASS   |
|   F32   |      I64      |      2^24      |        1         |  68.685 us |       1.04% |  68.736 us |       1.02% |   0.050 us |   0.07% |   PASS   |
|   F32   |      I64      |      2^28      |        1         | 972.431 us |       0.08% | 973.130 us |       0.09% |   0.699 us |   0.07% |   PASS   |
|   F64   |      I32      |      2^16      |        1         |   5.340 us |       3.47% |   5.434 us |       4.11% |   0.093 us |   1.75% |   PASS   |
|   F64   |      I32      |      2^20      |        1         |  13.932 us |       2.04% |  14.058 us |       2.26% |   0.126 us |   0.90% |   PASS   |
|   F64   |      I32      |      2^24      |        1         | 128.444 us |       0.61% | 128.647 us |       0.60% |   0.203 us |   0.16% |   PASS   |
|   F64   |      I32      |      2^28      |        1         |   1.930 ms |       0.04% |   1.932 ms |       0.04% |   2.511 us |   0.13% |   FAIL   |
|   F64   |      I64      |      2^16      |        1         |   5.472 us |       4.07% |   5.553 us |       4.79% |   0.081 us |   1.48% |   PASS   |
|   F64   |      I64      |      2^20      |        1         |  13.980 us |       2.09% |  14.026 us |       2.16% |   0.046 us |   0.33% |   PASS   |
|   F64   |      I64      |      2^24      |        1         | 128.508 us |       0.62% | 128.624 us |       0.61% |   0.116 us |   0.09% |   PASS   |
|   F64   |      I64      |      2^28      |        1         |   1.930 ms |       0.04% |   1.932 ms |       0.04% |   2.287 us |   0.12% |   FAIL   |
|  I128   |      I32      |      2^16      |        1         |   6.115 us |       4.30% |   6.116 us |       4.36% |   0.002 us |   0.03% |   PASS   |
|  I128   |      I32      |      2^20      |        1         |  21.890 us |       2.09% |  22.032 us |       2.18% |   0.143 us |   0.65% |   PASS   |
|  I128   |      I32      |      2^24      |        1         | 248.269 us |       0.35% | 247.838 us |       0.36% |  -0.430 us |  -0.17% |   PASS   |
|  I128   |      I32      |      2^28      |        1         |   3.861 ms |       0.04% |   3.851 ms |       0.04% |  -9.970 us |  -0.26% |   FAIL   |
|  I128   |      I64      |      2^16      |        1         |   5.901 us |       3.41% |   5.934 us |       3.28% |   0.033 us |   0.56% |   PASS   |
|  I128   |      I64      |      2^20      |        1         |  21.877 us |       2.14% |  22.058 us |       2.11% |   0.181 us |   0.83% |   PASS   |
|  I128   |      I64      |      2^24      |        1         | 248.336 us |       0.36% | 247.890 us |       0.36% |  -0.447 us |  -0.18% |   PASS   |
|  I128   |      I64      |      2^28      |        1         |   3.861 ms |       0.04% |   3.851 ms |       0.04% |  -9.888 us |  -0.26% |   FAIL   |
```